linkerd/detect/src/lib.rs

```diff
@@ -5,25 +5,27 @@ use linkerd_error::Error;
 use linkerd_io as io;
 use linkerd_stack::{layer, NewService};
 use std::{
+    fmt,
     future::Future,
     pin::Pin,
     task::{Context, Poll},
 };
 use tokio::time;
 use tower::util::ServiceExt;
-use tracing::{debug, trace};
+use tracing::{debug, info, trace};
 
 #[async_trait::async_trait]
-pub trait Detect: Clone + Send + Sync + 'static {
+pub trait Detect<I>: Clone + Send + Sync + 'static {
     type Protocol: Send;
 
-    async fn detect<I: io::AsyncRead + Send + Unpin + 'static>(
-        &self,
-        io: &mut I,
-        buf: &mut BytesMut,
-    ) -> Result<Option<Self::Protocol>, Error>;
+    async fn detect(&self, io: &mut I, buf: &mut BytesMut)
+        -> Result<Option<Self::Protocol>, Error>;
 }
 
+pub type Detected<P> = Result<Option<P>, DetectTimeout<P>>;
```

Comment:
nit/TIOLI: maybe `DetectResult` makes it clearer that this is a `Result` type alias? But that feels wordier so maybe this is better...I don't really care.